tensorflow_addons/optimizers/discriminative_layer_training.py

```diff
@@ -0,0 +1,255 @@
+                layers_with_lr_mult.append(sub_layer)
+                if verbose:
+                    logging.info("layer %s lr_mult : %f" % (sub_layer.name, lr_mult))
+
+        return layers_with_lr_mult
+
+    def _compute_params(self, var_list):
+        """helps compute params to provide a summary that aligns with model.summary()
+        """
+        return np.sum([np.prod(list(var.shape)) for var in var_list])
+
+    def _prepare_model(self, model, verbose=True):
+        """Prepares a built model for disc training
+        """
+
+        layers_with_lr_mult = self._check_for_lr_mult(model, verbose=verbose)
+        if len(layers_with_lr_mult) == 0:
+
+            logging.warning(
+                "Discriminative Layer Training requires an lr_mult attribute on at least one layer"
+            )
+
+            logging.warning(
+                "The assigned lr_mult must not be equal to 1. eg: model.layers[0].lr_mult = 0.01"
+            )
+
+        for layer in self._get_lowest_layers(model):
+            self._apply_lr_mult_to_var(layer)
+
+        vars_with_lr_mult = [
+            var for var in model.trainable_variables if var.lr_mult_value != 1.0
+        ]
+
+        if verbose:
+            logging.info(
+                "%i params of %i will learn at a different rate"
+                % (
+                    self._compute_params(vars_with_lr_mult),
+                    self._compute_params(model.trainable_variables),
+                )
+            )
+
+
+class DiscriminativeWrapper(tf.keras.optimizers.Optimizer):
+    @typechecked
+    def __init__(
+        self,
+        base_optimizer: object,
+        model: tf.keras.Model,
+        learning_rate: float,
+        verbose: bool = True,
+        name: str = "discrim_opt",
+        *args,
+        **kwargs
+    ):
+        """Discriminative Layer Training Wrapper
+
+            Discriminative layer training is a technique that applies different learning rates to
+            different layers in a model. Generally, a lower learning rate is applied to the
+            layers closest to the input and a higher learning rate is applied to layers closer
+            to the output. This method helps in transfer learning by quickly calibrating the head
+            of a model while preserving the useful weights in the main part of the model.
+
+            You should assign the lr_mult attribute to a layer. This will multiply the learning rate
+            used by the base optimizer for that layer.
+
+            This method creates a copy of the base optimizer for each unique learning rate multipler.
+
+            Performance is similar to using a single copy of the base optimizer as gradients are computed
+            only once and then passed on.
+
+            This optimizer does not support from_config or get_config. To try to preserve the state, you may
+            serialize the optimizers in the optimizer_group attribute in an instance of this class.
```

Comment:
I can look further into the serialization of this optimizer. At first, it seemed like a daunting task because it actually includes multiple nested optimizers, which I assume would each need to be serialized in some way.

Reply:
added warning on config and best attempt at get config